### PR TITLE
Add Redis table Engine/Function

### LINKED ISSUE #46146: Add `redis` table function and `Redis` table engine.
See #37865

### PR DESCRIPTION
**Pull request by JackyWoo**
This PR try to close #46146

### Changelog category (leave one):
- New Feature

Implements storage in the Redis.
Usage : ENGINE = Redis(host:port[, db_index[, password[, pool_size]]]) PRIMARY KEY(key)

In current design, redis table engine/function must have only one column as primary key who is redis key. And  the other columns will be serialized together as redis value.

Redis table engine/function now supports

1. select, insert, update, delete, truncate operations. 
2. equals, in predictes pushdown.

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
1. Add `TableFunctionRedis`
3. Add table engine Redis
4. Add `RedisCommon` which contains Redis related tools and types
5. Support `equals` and `in` filter push down into Redis

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

### Tests

- [x]  Integration tests

### CODE REVIEW

**Comment by robot-ch-test-poll3** [comment]
This is an automated comment for commit eddb9b9d12934b5fc8ae8d3337f07198e873b63b with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/50150/eddb9b9d12934b5fc8ae8d3337f07198e873b63b/ci_running.html)
The overall status of the commit is 🔴 failure

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>Style Check</td><td>Runs a set of checks to keep the code style clean. If some of tests failed, see the related log from the report</td><td>🔴 failure</td></tr>
</table>

**Comment by helifu** [comment]
Sorry, I'm interested in this feature, but could you please share the usage scenarious of it? Thanks in advance :)

**Comment by JackyWoo** [comment]
@helifu n my production scenario, there is a need for flexibility in querying as well as a requirement for high query throughput. I want to promote query throughput Redis.

Usage:

1. `Insert-into-select` data to Redis and query thought Redis directly.
2. Insert data to Redis and query throught ClickHouse.

**Comment by helifu** [comment]
> 

Ah, that's cool. But it seems that redis can't offer high throughput since it's a key-value storage IIRC.

**Comment by JackyWoo** [comment]
I will work in key based query scenarios.

**Comment by JackyWoo** [comment]
@pufit Please review it when you have a chance. There are some test errors but maybe unrelated with the PR.

**Comment by JackyWoo** [comment]
@pufit May I ask when you have a chance to review th PR?

File: src/Storages/StorageRedis.cpp [hunk 1]
```diff
@@ -0,0 +1,579 @@
...
+            return {};
+        }
+
+        const auto & key_column_type = sample_block.getByName(storage.getPrimaryKey().at(0)).type;
+        auto raw_keys = serializeKeysToRawString(it, end, key_column_type, max_block_size);
+
+        return storage.getBySerializedKeys(raw_keys, nullptr);
+    }
+
+    /// TODO scan may get duplicated keys
```
Comment (pufit) [inline src/Storages/StorageRedis.cpp:90]: Can you provide more info about this issue?

File: src/Storages/StorageRedis.cpp [hunk 2]
```diff
@@ -0,0 +1,579 @@
...
+
+        RedisArray scan_keys;
+        RedisIterator next_iterator;
+
+        std::tie(next_iterator, scan_keys) = storage.scan(iterator == -1 ? 0 : iterator, pattern, max_block_size);
+        iterator = next_iterator;
+
+        /// redis scan can return nothing
+        if (scan_keys.isNull() || scan_keys.size() == 0)
+            return generateFullScan();
```
Comment (pufit) [inline src/Storages/StorageRedis.cpp:105]: This looks dangerous. Do we know why Redis can return nothing?

**Review by pufit — APPROVED** [review]
Besides some small questions, LGTM!

**Comment by JackyWoo** [comment]
@pufit for the issue duplicated keys.

To perfectly resolve it, we need additional rehashing stat info from Redis, I submit an issue to Redis for it.  And now we have two ways:

1. Just leave it and add notifications in docs. For duplication is a vary rare case and we can fix it when Redis can provide rehasing stat info. 
2. Keep two scan results in memory and filter them every scan.

I prefer the second one, but any way I will take your opinion?

**Comment by pufit** [comment]
> @pufit for the issue duplicated keys.
> 
> To perfectly resolve it, we need additional rehashing stat info from Redis, I submit an issue to Redis for it. And now we have two ways:
> 
> 1. Just leave it and add notifications in docs. For duplication is a vary rare case and we can fix it when Redis can provide rehasing stat info.
> 2. Keep two scan results in memory and filter them every scan.
> 
> I prefer the second one, but any way I will take your opinion?

Well, actually, the first option looks much more straightforward. All duplications can be handled on ClickHouse side (using `DISTINCT`, for example). And we can improve this behavior in the future if we consider it necessary.

**Comment by JackyWoo** [comment]
@pufit Thank you, I will adjust the docs.

File: docs/en/engines/table-engines/integrations/redis.md [hunk 3]
```diff
@@ -110,3 +110,10 @@ Flush Redis db asynchronously. Also `Truncate` support SYNC mode.
 ```sql
 TRUNCATE TABLE redis_table SYNC;
 ```
+
+
+## Limitations {#limitations}
+
+Redis engine also support scanning query, such as `where k > xx`, but it has some limitations:
```
Comment (pufit) [inline docs/en/engines/table-engines/integrations/redis.md:117]:
```suggestion
Redis engine also supports scanning queries, such as `where k > xx`, but it has some limitations:
```

File: docs/en/engines/table-engines/integrations/redis.md [hunk 4]
```diff
@@ -110,3 +110,10 @@ Flush Redis db asynchronously. Also `Truncate` support SYNC mode.
...
+1. Scanning query may produce some duplicated keys in a very rare case when it is rehashing, details see [Redis Scan](https://github.com/redis/redis/blob/e4d183afd33e0b2e6e8d1c79a832f678a04a7886/src/dict.c#L1186-L1269)
```
Comment (pufit) [inline docs/en/engines/table-engines/integrations/redis.md:118]:
```suggestion
1. Scanning query may produce some duplicated keys in a very rare case when it is rehashing. See details in [Redis Scan](https://github.com/redis/redis/blob/e4d183afd33e0b2e6e8d1c79a832f678a04a7886/src/dict.c#L1186-L1269)
```

File: docs/en/engines/table-engines/integrations/redis.md [hunk 5]
```diff
@@ -110,3 +110,10 @@ Flush Redis db asynchronously. Also `Truncate` support SYNC mode.
...
+2. During the scanning keys could be created and deleted, so the resulting dataset can not represent a valid point in time.
```
Comment (pufit) [inline docs/en/engines/table-engines/integrations/redis.md:119]:
```suggestion
2. During the scanning, keys could be created and deleted, so the resulting dataset can not represent a valid point in time.
```

**Review by pufit — APPROVED** [review]

**Comment by pufit** [comment]
Failed tests:
`test_async_query_sending` - https://github.com/ClickHouse/ClickHouse/pull/50480
`02479_mysql_connect_to_self` - https://github.com/ClickHouse/ClickHouse/issues/50911

**Comment by gingerwizard** [comment]
@JackyWoo late to see this but started testing it. I see you always use MGET command and never HMGET - the latter feels more natural for table data. Anyway we could support this as MGET means the values must be a string?

**Comment by stm32repo** [comment]
Hello. Why does the Redis engine add the character \x01 at the beginning when reading and inserting strings

**Pull request merged**